flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/MapRangePartitioner.java

```diff
@@ -0,0 +1,288 @@
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.concurrent.ThreadLocalRandom;
+import org.apache.flink.api.common.functions.Partitioner;
+import org.apache.flink.table.data.RowData;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SortKey;
+import org.apache.iceberg.SortOrder;
+import org.apache.iceberg.SortOrderComparators;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.flink.FlinkSchemaUtil;
+import org.apache.iceberg.flink.RowDataWrapper;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.ArrayUtil;
+import org.apache.iceberg.util.Pair;
+
+/**
+ * Internal partitioner implementation that supports MapDataStatistics, which is typically used for
+ * low-cardinality use cases. While MapDataStatistics can keep accurate counters, it can't be used
+ * for high-cardinality use cases. Otherwise, the memory footprint is too high.
+ */
+class MapRangePartitioner implements Partitioner<RowData> {
+  private final RowDataWrapper rowDataWrapper;
+  private final SortKey sortKey;
+  private final Comparator<StructLike> comparator;
+  private final Map<SortKey, Long> mapStatistics;
+  private final double closeFileCostInWeightPercentage;
+
+  // lazily computed due to the need of numPartitions
+  private Map<SortKey, KeyAssignment> assignment;
+  private NavigableMap<SortKey, Long> sortedStatsWithCloseFileCost;
+
+  MapRangePartitioner(
+      Schema schema,
+      SortOrder sortOrder,
+      MapDataStatistics dataStatistics,
+      double closeFileCostInWeightPercentage) {
+    this.rowDataWrapper = new RowDataWrapper(FlinkSchemaUtil.convert(schema), schema.asStruct());
+    this.sortKey = new SortKey(schema, sortOrder);
+    this.comparator = SortOrderComparators.forSchema(schema, sortOrder);
+    this.mapStatistics = dataStatistics.statistics();
+    this.closeFileCostInWeightPercentage = closeFileCostInWeightPercentage;
+  }
+
+  @Override
+  public int partition(RowData row, int numPartitions) {
+    // assignment table can only be built lazily when first referenced here,
+    // because number of partitions (downstream subtasks) is needed
+    Map<SortKey, KeyAssignment> assignmentMap = assignment(numPartitions);
+    // reuse the sortKey and rowDataWrapper
+    sortKey.wrap(rowDataWrapper.wrap(row));
+    KeyAssignment keyAssignment = assignmentMap.get(sortKey);
+    if (keyAssignment == null) {
+      // haven't learned about the key before. fall back to random selection.
+      return ThreadLocalRandom.current().nextInt(numPartitions);
```

Comment:
Random, or hash is better?

I think random is prone to OOM too, as we could end up with too many open files. Hash might be slow, but after the first checkpoint it could work.
WDYT?

Reply:
Great point and question.

I see multiple options here and we have to choose the least evil.

1. random as implemented current, which is prone to OOM as you pointed out
2. `sortKey.hashCode() % numPartitions`, which is probably what you meant by hash. this can also
3. direct forward. subtask x -> subtask x

I agree that hash is better than random on avoiding OOM, even though hash is prone to unbalanced traffic distribution. all traffic for the new key go to a single subtask, which can lead to back pressure. back pressure is probably better than OOM.

I am wondering if #3 direct forwarding is better than #2 hash. We will have more files for the new key. the benefit is that the traffic for the new key is evenly spread across downstream subtasks.

Reply:
Let's discuss different scenarios

1. only one new sort key

agree that direct forward and random seems be identical in this case. Essentially, every downstream task needs to write one more data file for the new key. one more file shouldn't lead to OOM normally.

In this case, direct forward or random is probably better.

2. many new sort key (say K)

in this case, hash can be more desirable to distribute K shards to N downstream subtasks. otherwise, with random/direct forward, each subtask will need to process K new data files. but simple keyBy/hash tends to produce unbalanced distribution as mentioned by [PR 4228](https://github.com/apache/iceberg/pull/4228).

> high cardinality columns 

that needs to be handled differently. we would need to implement sketch sampling algorithm to support high cardinality columns.

Considering map statistics is only targeted for low-cardinality scenario, maybe random/round-robin/direct-forward could be a better default behavior?

Reply:
2. with random/direct forward could lead to OOM

Reply:
for case 2 (many new unknown sort keys), it is less likely to happen for low-cardinality scenario. basically, I am trying to make a case for the default behavior to be `no shuffling/clustering`. Mixing hash shuffling for unknown keys with range shuffling for know keys seems a bit odd combination.

Reply:
I leave this to you.
My only concern is that one of the goals was to keep the number of files to small, and this will likely create `numPartitions` number of files. If the new keys cardinality is high, then it is good, but if the new keys cardinality is low, then this is an issue.
But if we need this config later, we can add whenever the requirement pops up.

Reply:
>  one of the goals was to keep the number of files to small

that is correct. Just want to make sure we are on the same page. We are not talking about new sort columns. We are talking about new unknown values from the SortKey. E.g., we are sorting on `event_type` column and there is a new event_type value added to the data stream. because we don't have traffic distribution/weight for this new `event_type` value, we will not shuffle the records with the new `event_type` value. After two checkpoint cycles and the learned weight for this new value, range shuffle is applied. yes, for two checkpoint cycles, we can have `numPartitions` more data files for the new value. this is like we don't enable range shuffle.

Reply:
I absolutely agree, that whatever solution we choose, we are better than any of the original solutions!

Reply:
I will go with round robin `AtomicLong.getAndIncrement() % N` then. it might be a tiny bit faster than `ThreadLocalRandom.current.nextInt(N)`